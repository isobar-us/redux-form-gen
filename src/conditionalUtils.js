// @flow
import get from 'lodash/get';
import has from 'lodash/has';
import hasIn from 'lodash/hasIn';
import includes from 'lodash/includes';
import omit from 'lodash/omit';
import isEqual from 'lodash/isEqual';
import isNumber from 'lodash/isNumber';
import isNil from 'lodash/isNil';
import isString from 'lodash/isString';
import {isNilOrEmpty, isFieldFilled, isFieldValid} from './validators';

import type {FieldType} from './types';
import type {ConditionalOperators, EvalCondOptions, ConditionalObject} from './conditionalUtils.types';

/*
  Helpers
 */

// used for render lifecycle where we don't care about conditionalValid
export const isCondField = (field: FieldType) =>
  has(field, 'conditionalVisible') || has(field, 'conditionalRequired') || has(field, 'conditionalDisabled');

const getCondValueKey = ({cond, valueKey}) => (has(cond, 'questionId') ? cond.questionId : valueKey);

const getOperator = (options, key) => {
  const {customOperators} = options;
  const operators = {...ops, ...(customOperators || {})};
  const op = get(operators, key);
  if (isNil(op)) {
    console.error(`[FormGenerator] Unknown conditional operator "${key}". Condition:`, options.cond);
  }
  return op;
};

/*
  Conditional Operators
 */
const ops: ConditionalOperators = {
  equals: ({value, param}) => isEqual(value, param),
  // cond
  and: ({value, param, ...options}) =>
    param.reduce((and, subCond) => and && evalCond({...options, cond: subCond}), true),
  or: ({value, param, ...options}) => param.reduce((or, subCond) => or || evalCond({...options, cond: subCond}), false),
  not: ({value, param, ...options}) => !evalCond({...options, cond: param}),
  cond: ({value, param, ...options}) => evalCond({...options, cond: param}),
  compare: (options) => {
    const {value, param, data} = options;

    const conds = Object.keys(omit(param, ['questionId']));
    return conds.length > 0
      ? conds.reduce((result, key) => {
          // will AND all the cond props
          const operator = getOperator(options, key);
          if (isNil(operator)) {
            return true;
          }
          const remoteQuestionId = get(param, key);
          const compareParam = get(
            data,
            getCondValueKey({
              ...options,
              cond: {
                ...(remoteQuestionId && {questionId: remoteQuestionId})
              }
            })
          );
          return result && operator({...options, value, param: compareParam});
        }, true)
      : true;
  },
  // generator hooks
  filled: ({value, param, ...options}) => {
    let fieldFilled = true;
    if (has(options, 'lookupTable')) {
      const {lookupTable} = options;
      const field = get(lookupTable, getCondValueKey(options));
      if (!isNil(field)) {
        // if field is defined in the lookupTable, run isFieldFilled()
        fieldFilled = isFieldFilled({...options, field, fieldOptions: null});
      } else {
        // if field is not defined in the lookupTable, just fall back to isNilOrEmpty
        fieldFilled = !isNilOrEmpty(value);
      }
    } else {
      console.warn('[Form Generator] attempted to check `filled` without a lookupTable. Condition:', options.cond); // TODO finalise this warning
      fieldFilled = !isNilOrEmpty(value);
    }
    return param === true ? fieldFilled : !fieldFilled;
  },
  // TODO experimental operator to check if this or another field is valid...
  // valid: ({value, param, ...options}) => {
  //   const {cond, lookupTable} = options;
  //   let field = options.field;
  //
  //   if (has(cond, 'questionId')) {
  //     if (has(options, 'lookupTable')) {
  //       field = get(lookupTable, cond.questionId);
  //     } else {
  //       console.warn('[Form Generator] attempted to check `valid` without a lookupTable. Condition:', options.cond); // TODO finalise this warning
  //     }
  //   }
  //
  //   const fieldValid = isFieldValid({
  //     ...options,
  //     field,
  //     fieldOptions: null,
  //   });
  //   return param === true ? fieldValid : !fieldValid;
  // },
  // lodash
  includes: ({value, param}) => includes(value, param),
  // comparison
  greaterThan: ({value, param}) => {
    if (isNilOrEmpty(value)) {
      return false;
    }
    const parsedValue = isNumber(value) ? value : parseFloat(value);
    return parsedValue > param;
  },
  lessThan: ({value, param}) => {
    if (isNilOrEmpty(value)) {
      return false;
    }
    const parsedValue = isNumber(value) ? value : parseFloat(value);
    return parsedValue < param;
  },
  greaterThanEqual: ({value, param}) => {
    if (isNilOrEmpty(value)) {
      return false;
    }
    const parsedValue = isNumber(value) ? value : parseFloat(value);
    return parsedValue >= param;
  },
  lessThanEqual: ({value, param}) => {
    if (isNilOrEmpty(value)) {
      return false;
    }
    const parsedValue = isNumber(value) ? value : parseFloat(value);
    return parsedValue <= param;
  },
  // length
  length: ({value, param}) => (isNil(value) ? false : hasIn(value, 'length') ? value.length === param : true),
  minLength: ({value, param}) => (isNil(value) ? false : hasIn(value, 'length') ? value.length >= param : true),
  maxLength: ({value, param}) => (hasIn(value, 'length') ? value.length <= param : true),
  // regex
  pobox: ({value, param}) => (param && isString(value) ? /p.*o.* box/gi.test(value) : true),
  email: ({value, param}) => (param && isString(value) ? /\S+@\S+\.\S+/.test(value) : true),
  regex: ({value, param}) => (isString(value) ? RegExp(param).test(value) : true)
};

/*
  Conditional Operator Aliases
 */
ops.gt = ops.greaterThan;
ops.lt = ops.lessThan;
ops.min = ops.greaterThanEqual;
ops.max = ops.lessThanEqual;
ops.value = ops.equals;

const defaultElseHandler = (options) => ops.filled({...options, param: true});

const defaultValueKey = '_value';

export const evalCond = (options: EvalCondOptions) => {
  options = {
    elseHandler: defaultElseHandler,
    valueKey: defaultValueKey,
    ...options
  };
  const {cond, data, elseHandler} = options;

  // TODO add support for getFieldPath() when doing get(data, path) ?
  const value = get(data, getCondValueKey(options));
  const conds = Object.keys(omit(cond, 'questionId'));

  return conds.length > 0
    ? conds.reduce((result, key) => {
        // will AND all the cond props
        const operator = getOperator(options, key);
        if (isNil(operator)) {
          return true;
        }
        const param = get(cond, key);
        return result && operator({...options, value, param});
      }, true)
    : elseHandler({...options, value});
};

const condValidElseHandler = (options) => {
  const {cond, lookupTable} = options;
  let field = options.field;
  let fieldOptions = null;

  if (has(cond, 'questionId')) {
    if (has(options, 'lookupTable')) {
      field = get(lookupTable, cond.questionId);
      fieldOptions = null;
    } else {
      console.warn('[Form Generator] attempted to check `valid` without a lookupTable. Condition:', options.cond); // TODO finalise this warning
    }
  }

  return isFieldValid({
    ...options,
    field,
    fieldOptions
  });

  // TODO hook into experimental `valid` operator when ready
  // return ops.valid({...options, param: true});
};
export const evalCondValid = (options: EvalCondOptions) => {
  const {data} = options;
  const value = get(data, getCondValueKey(options));

  // can not validate if the field is empty, so return true
  if (isNil(value) || value === '') {
    return true;
  }

  return evalCond({
    elseHandler: condValidElseHandler,
    ...options
  });
};

export default {
  evalCond,
  evalCondValid
};

// TODO consider renaming to getCondDependentFieldNames or something similar
export const condDependentFields = (cond: ConditionalObject) => {
  if (has(cond, 'questionId')) {
    return [cond.questionId];
  } else if (has(cond, 'and') && Array.isArray(cond.and)) {
    return cond.and.reduce((a, subCond) => a.concat(condDependentFields(subCond)), []);
  } else if (has(cond, 'or') && Array.isArray(cond.or)) {
    return cond.or.reduce((a, subCond) => a.concat(condDependentFields(subCond)), []);
  } else if (has(cond, 'not') && cond.not) {
    return condDependentFields(cond.not);
  }
  return [];
};
