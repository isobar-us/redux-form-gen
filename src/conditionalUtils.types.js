import type {FieldType} from './types';
export type ConditionalObject = {
  questionId: string,
  equals?: mixed,
  // cond
  and?: Array<ConditionalObject>,
  or?: Array<ConditionalObject>,
  not?: ConditionalObject,
  // lodash
  filled?: boolean,
  includes?: mixed | Array<mixed>,
  // comparison
  greaterThan?: number,
  lessThan?: number,
  greaterThanEqual?: number,
  lessThanEqual?: number,
  // length
  length: number,
  minLength: number,
  maxLength: number,
  // regex
  pobox: boolean,
  email: boolean,
  regex: boolean
};

export type ConditionalDependency = {
  questionId: string,
  globalScope?: Boolean
};

export type FieldDependencyOptions = {
  field: FieldType,
  parentPath: String
};

export type EvalCondOptions = {
  elseHandler?: Function,
  valueKey?: string,
  cond: ConditionalObject,
  customOperators: CustomOperators // eslint-disable-line no-use-before-define
};

export type CustomOperators = {
  [key: string]: {(options: EvalCondOptions): boolean}
};

export type ConditionalOperatorOptions = {
  value: mixed,
  param: mixed,
  ...EvalCondOptions
};

type ConditionalOperatorNumberParam = {
  (options: {...ConditionalOperatorOptions, value: number, param: number}): boolean
};
type ConditionalOperatorObjectParam = {
  (options: {...ConditionalOperatorOptions, value: Object, param: boolean}): boolean
};

type ConditionalOperator = {(options: ConditionalOperatorOptions): boolean};

export type ConditionalOperators = {
  equals: ConditionalOperator,
  and: {(options: {...ConditionalOperatorOptions, param: Array<ConditionalObject>}): boolean},
  or: {(options: {...ConditionalOperatorOptions, param: Array<ConditionalObject>}): boolean},
  not: {(options: {...ConditionalOperatorOptions, param: ConditionalObject}): boolean},
  filled: ConditionalOperator,
  includes: {
    (options: {...ConditionalOperatorOptions, value: Array<mixed>, param: mixed}): boolean
  },
  greaterThan: ConditionalOperatorNumberParam,
  lessThan: ConditionalOperatorNumberParam,
  greaterThanEqual: ConditionalOperatorNumberParam,
  lessThanEqual: ConditionalOperatorNumberParam,
  length: ConditionalOperatorObjectParam,

  gt?: ConditionalOperatorNumberParam,
  lt?: ConditionalOperatorNumberParam,
  min?: ConditionalOperatorNumberParam,
  max?: ConditionalOperatorNumberParam,
  value?: ConditionalOperator
};
