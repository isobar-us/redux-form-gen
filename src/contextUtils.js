// @flow
import React from 'react';
import PropTypes from 'prop-types';

import createReactContext from 'create-react-context';
export const GenContext = createReactContext({wasGenerated: false});

export const consumeGenContext = (Component: Function) => (props: mixed) => (
  <GenContext.Consumer>
    {(gen) => <Component {...props} gen={gen} /> }
  </GenContext.Consumer>
);

// inspired by
// https://medium.com/react-ecosystem/how-to-handle-react-context-a7592dfdcbc

// const provideContext = (childContextTypes, getChildContext) => (WrappedComponent) => {
//   class ContextProvider extends Component {
//     static childContextTypes = childContextTypes;
//     getChildContext = () => getChildContext(this.props);
//
//     render() {
//       return <WrappedComponent {...this.props} />;
//     }
//   }
//   return ContextProvider;
// };

const consumeContext = (contextTypes) => (Component: Function) => {
  /* The context is passed as props. This way the component is
   completely decoupled from the context API.
  */
  const ContextConsumer = (props: mixed, context: mixed) => <Component {...props} {...context} />;
  ContextConsumer.contextTypes = contextTypes;
  return ContextConsumer;
};

export const reduxFormContext = {
  _reduxForm: PropTypes.object
};

export const consumeReduxFormContext = consumeContext(reduxFormContext);
