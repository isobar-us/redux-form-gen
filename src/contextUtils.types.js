import type {Props as FormGeneratorProps} from './FormGenerator.types';

export type GenContextProps = {
  ...FormGeneratorProps,
  getCachedValue: Function,
  setCachedValue: Function
};
