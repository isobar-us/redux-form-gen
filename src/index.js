// @flow
import FormGenerator from './FormGenerator';

import defaultFieldTypes, {getFieldOptions, genericFieldProps} from './defaultFieldTypes';
import GenericRequiredLabel from './defaultFieldTypes/components/GenericRequiredLabel';
import RequiredIndicator from './defaultFieldTypes/components/RequiredIndicator';
import injectGenProps from './injectGenProps';

import {
  isSectionEmpty,
  isSectionEmptyIterator,
  isSectionFilled,
  isSectionFilledIterator,
  getSectionErrors,
  getSectionErrorsIterator,
  isNilOrEmpty,
  getFieldPath,
  isFieldHidden,
  isFieldVisible,
  isFieldRequired,
  isFieldDisabled,
  isFieldEmpty,
  isFieldFilled,
  isFieldValid,
  mapFieldChildren
} from './validators';

import {getDefaultValues, buildLookupTable} from './utils';

import {consumeGenContext, GenContext} from './contextUtils';

import {evalCond, evalCondValid} from './conditionalUtils';

import GenField from './GenField';
import GenWrapper from './GenWrapper';

export type {FieldOptions, CustomFieldTypes} from './types';

export {
  FormGenerator,
  // defaultFieldTypes
  defaultFieldTypes,
  getFieldOptions,
  RequiredIndicator,
  GenericRequiredLabel,
  genericFieldProps,
  injectGenProps,
  // validators
  isSectionEmpty,
  isSectionEmptyIterator,
  isSectionFilled,
  isSectionFilledIterator,
  getSectionErrors,
  getSectionErrorsIterator,
  isNilOrEmpty,
  // validator utils
  getFieldPath,
  isFieldHidden,
  isFieldVisible,
  isFieldRequired,
  isFieldDisabled,
  isFieldEmpty,
  isFieldFilled,
  isFieldValid,
  mapFieldChildren,
  // utils
  getDefaultValues,
  buildLookupTable,
  // contextUtils
  consumeGenContext,
  GenContext,
  // conditionalUtils
  evalCond,
  evalCondValid,
  // internals
  GenField,
  GenWrapper
};
