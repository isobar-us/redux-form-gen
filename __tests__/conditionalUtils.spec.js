import {evalCond, evalCondValid, condDependentFields} from '../src/conditionalUtils';

import {buildLookupTable} from '../src/utils';

let consoleErrorSpy;
let consoleWarnSpy;
const startErrorSupression = () => {
  consoleErrorSpy = jest.spyOn(console, 'error');
  consoleErrorSpy.mockImplementation(() => {}); // suppress error log
  consoleWarnSpy = jest.spyOn(console, 'warn');
  consoleWarnSpy.mockImplementation(() => {}); // suppress warn log
};

const stopErrorSuppression = () => {
  consoleErrorSpy.mockReset();
  consoleErrorSpy.mockRestore();
  consoleWarnSpy.mockReset();
  consoleWarnSpy.mockRestore();
};

describe('evalCond()', () => {
  const fooFields = [
    {
      type: 'text',
      questionId: 'foo'
    }
  ];

  const fooLookupTable = buildLookupTable({
    fields: fooFields
    // customFieldTypes
  });

  describe('default cond (filled)', () => {
    const lookupTable = fooLookupTable;
    it('should return true if remote questionId filled out', () => {
      expect(
        evalCond({
          cond: {questionId: 'foo'},
          data: {foo: 'bar'},
          lookupTable
        })
      ).toBe(true);
    });

    it('should return false if remote questionId is empty', () => {
      expect(
        evalCond({
          cond: {questionId: 'foo'},
          data: {foo: ''},
          lookupTable
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {questionId: 'foo'},
          data: {foo: null},
          lookupTable
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {questionId: 'foo'},
          data: {},
          lookupTable
        })
      ).toBe(false);
    });
  });

  describe('filled', () => {
    const lookupTable = fooLookupTable;
    it('should return true if remote questionId filled out', () => {
      expect(
        evalCond({
          cond: {questionId: 'foo', filled: true},
          data: {foo: 'bar'},
          lookupTable
        })
      ).toBe(true);
    });

    it('should return false if remote questionId is empty', () => {
      expect(
        evalCond({
          cond: {questionId: 'foo', filled: true},
          data: {foo: ''},
          lookupTable
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {questionId: 'foo', filled: true},
          data: {foo: null},
          lookupTable
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {questionId: 'foo', filled: true},
          data: {},
          lookupTable
        })
      ).toBe(false);
    });

    it('should return true if remote questionId is empty', () => {
      expect(
        evalCond({
          cond: {questionId: 'foo', filled: false},
          data: {foo: ''},
          lookupTable
        })
      ).toBe(true);

      expect(
        evalCond({
          cond: {questionId: 'foo', filled: false},
          data: {foo: null},
          lookupTable
        })
      ).toBe(true);

      expect(
        evalCond({
          cond: {questionId: 'foo', filled: false},
          data: {},
          lookupTable
        })
      ).toBe(true);
    });

    // it('should return true if remote questionId is empty', () => {
    //   expect(evalCond({
    //     cond: {equals: 'foo', filled: 'other'},
    //     data: {foo: ''}
    //   })).toBe(true);
    // });
  });

  describe('valueKey', () => {
    const lookupTable = fooLookupTable;
    it('should return true if parent filled out', () => {
      expect(
        evalCond({
          cond: {equals: 'bar'},
          data: {foo: 'bar'},
          valueKey: 'foo'
        })
      ).toBe(true);
    });

    it('should return false if parent is empty', () => {
      expect(
        evalCond({
          cond: {questionId: 'foo'},
          data: {foo: ''},
          valueKey: 'foo',
          lookupTable
        })
      ).toBe(false);
    });
  });

  describe('not', () => {
    it('should return false if foo === "bar"', () => {
      expect(
        evalCond({
          cond: {not: {questionId: 'foo', equals: 'bar'}},
          data: {foo: 'bar'}
        })
      ).toBe(false);
    });

    it('should return true if foo !== "bar"', () => {
      expect(
        evalCond({
          cond: {not: {questionId: 'foo', equals: 'bar'}},
          data: {foo: 'baz'}
        })
      ).toBe(true);
    });
  });

  describe('includes', () => {
    it('should return true if foo includes "bar"', () => {
      expect(
        evalCond({
          cond: {questionId: 'foo', includes: 'bar'},
          data: {foo: ['bar']}
        })
      ).toBe(true);
    });

    it('should return false if foo does not include "bar"', () => {
      expect(
        evalCond({
          cond: {questionId: 'foo', includes: 'bar'},
          data: {foo: ['baz']}
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {questionId: 'foo', includes: 'bar'},
          data: {}
        })
      ).toBe(false);
    });
  });

  describe('lessThan', () => {
    it('should return true if foo less than 5', () => {
      expect(
        evalCond({
          cond: {questionId: 'foo', lessThan: 5},
          data: {foo: 2}
        })
      ).toBe(true);

      expect(
        evalCond({
          cond: {questionId: 'foo', lessThan: 5},
          data: {foo: '2'}
        })
      ).toBe(true);

      expect(
        evalCond({
          cond: {questionId: 'foo', lessThan: 5},
          data: {} // nil will result in false
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {questionId: 'foo', lessThan: 5},
          data: {foo: ''} // empty string will result in false
        })
      ).toBe(false);
    });

    it('should return false if foo is not less than 5', () => {
      expect(
        evalCond({
          cond: {questionId: 'foo', lessThan: 5},
          data: {foo: 5}
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {questionId: 'foo', lessThan: 5},
          data: {foo: '5'}
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {questionId: 'foo', lessThan: -5},
          data: {} // nil will result in false
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {questionId: 'foo', lessThan: -5},
          data: {foo: ''} // empty string will result in false
        })
      ).toBe(false);
    });
  });

  describe('greaterThan', () => {
    it('should return true if foo greater than 5', () => {
      expect(
        evalCond({
          cond: {questionId: 'foo', greaterThan: 5},
          data: {foo: 7}
        })
      ).toBe(true);

      expect(
        evalCond({
          cond: {questionId: 'foo', greaterThan: 5},
          data: {foo: '7'}
        })
      ).toBe(true);

      expect(
        evalCond({
          cond: {questionId: 'foo', greaterThan: -5},
          data: {} // nil will result in false
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {questionId: 'foo', greaterThan: -5},
          data: {foo: ''} // empty string will result in false
        })
      ).toBe(false);
    });

    it('should return false if foo is not greater than 5', () => {
      expect(
        evalCond({
          cond: {questionId: 'foo', greaterThan: 5},
          data: {foo: 5}
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {questionId: 'foo', greaterThan: 5},
          data: {foo: '5'}
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {questionId: 'foo', greaterThan: 5},
          data: {} // nil will result in false
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {questionId: 'foo', greaterThan: 5},
          data: {foo: ''} // empty string will result in false
        })
      ).toBe(false);
    });
  });

  describe('lessThanEqual', () => {
    it('should return true if foo less than 5', () => {
      expect(
        evalCond({
          cond: {questionId: 'foo', lessThanEqual: 5},
          data: {foo: 2}
        })
      ).toBe(true);

      expect(
        evalCond({
          cond: {questionId: 'foo', lessThanEqual: 5},
          data: {foo: '2'}
        })
      ).toBe(true);

      expect(
        evalCond({
          cond: {questionId: 'foo', lessThanEqual: 5},
          data: {} // nil will result in false
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {questionId: 'foo', lessThanEqual: 5},
          data: {foo: ''} // empty string will result in false
        })
      ).toBe(false);
    });

    it('should return false if foo is not less than 5', () => {
      expect(
        evalCond({
          cond: {questionId: 'foo', lessThanEqual: 5},
          data: {foo: 6}
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {questionId: 'foo', lessThanEqual: 5},
          data: {foo: '6'}
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {questionId: 'foo', lessThanEqual: -5},
          data: {} // nil will result in false
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {questionId: 'foo', lessThanEqual: -5},
          data: {foo: ''} // empty string will result in false
        })
      ).toBe(false);
    });
  });

  describe('greaterThanEqual', () => {
    it('should return true if foo greater than 5', () => {
      expect(
        evalCond({
          cond: {questionId: 'foo', greaterThanEqual: 5},
          data: {foo: 7}
        })
      ).toBe(true);

      expect(
        evalCond({
          cond: {questionId: 'foo', greaterThanEqual: 5},
          data: {foo: 5}
        })
      ).toBe(true);

      expect(
        evalCond({
          cond: {questionId: 'foo', greaterThanEqual: 5},
          data: {foo: '7'}
        })
      ).toBe(true);

      expect(
        evalCond({
          cond: {questionId: 'foo', greaterThanEqual: 5},
          data: {foo: '5'}
        })
      ).toBe(true);

      expect(
        evalCond({
          cond: {questionId: 'foo', greaterThanEqual: -5},
          data: {} // nil will result in false
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {questionId: 'foo', greaterThanEqual: -5},
          data: {foo: ''} // empty string will result in false
        })
      ).toBe(false);
    });

    it('should return false if foo is not greater than 5', () => {
      expect(
        evalCond({
          cond: {questionId: 'foo', greaterThanEqual: 5},
          data: {foo: 4}
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {questionId: 'foo', greaterThanEqual: 5},
          data: {foo: '4'}
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {questionId: 'foo', greaterThanEqual: 5},
          data: {} // nil will result in false
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {questionId: 'foo', greaterThanEqual: 5},
          data: {foo: ''} // empty string will result in false
        })
      ).toBe(false);
    });
  });

  describe('and', () => {
    it('should return true if both conds are true', () => {
      expect(
        evalCond({
          cond: {
            and: [{questionId: 'foo', gt: 2}, {questionId: 'foo', lt: 7}]
          },
          data: {foo: 5}
        })
      ).toBe(true);
    });

    it('should return false if at least one cond is false', () => {
      expect(
        evalCond({
          cond: {
            and: [{questionId: 'foo', gt: 2}, {questionId: 'foo', lt: 4}]
          },
          data: {foo: 5}
        })
      ).toBe(false);
    });
  });

  describe('or', () => {
    it('should return true if at least one cond is true', () => {
      expect(
        evalCond({
          cond: {
            or: [{questionId: 'foo', gt: 2}, {questionId: 'foo', lt: 4}]
          },
          data: {foo: 5}
        })
      ).toBe(true);
    });

    it('should return false if both conds are false', () => {
      expect(
        evalCond({
          cond: {
            or: [{questionId: 'foo', equals: 6}, {questionId: 'foo', lt: 4}]
          },
          data: {foo: 5}
        })
      ).toBe(false);
    });
  });

  describe('length', () => {
    it('should return true if at length = 3', () => {
      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            length: 3
          },
          data: {foo: 'abc'}
        })
      ).toBe(true);
    });

    it('should return false if at length != 3', () => {
      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            length: 3
          },
          data: {}
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            length: 3
          },
          data: {foo: null}
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            length: 3
          },
          data: {foo: ''}
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            length: 3
          },
          data: {foo: 's'}
        })
      ).toBe(false);
    });
  });

  describe('minLength', () => {
    it('should return true if at length >= 3', () => {
      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            minLength: 3
          },
          data: {foo: 'abcdef'}
        })
      ).toBe(true);
    });

    it('should return false if at length < 3', () => {
      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            minLength: 3
          },
          data: {}
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            minLength: 3
          },
          data: {foo: null}
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            minLength: 3
          },
          data: {foo: ''}
        })
      ).toBe(false);

      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            minLength: 3
          },
          data: {foo: 'ab'}
        })
      ).toBe(false);
    });
  });

  describe('maxLength', () => {
    it('should return true if at length <= 3', () => {
      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            maxLength: 3
          },
          data: {}
        })
      ).toBe(true);

      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            maxLength: 3
          },
          data: {foo: null}
        })
      ).toBe(true);

      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            maxLength: 3
          },
          data: {foo: ''}
        })
      ).toBe(true);

      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            maxLength: 3
          },
          data: {foo: 'abc'}
        })
      ).toBe(true);
    });

    it('should return false if at length > 3', () => {
      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            maxLength: 3
          },
          data: {foo: 'abcdef'}
        })
      ).toBe(false);
    });
  });

  describe('email', () => {
    it('should return true if nil', () => {
      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            email: true
          },
          data: {}
        })
      ).toBe(true);

      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            regex: true
          },
          data: {foo: null}
        })
      ).toBe(true);
    });

    it('should return true if is an email', () => {
      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            email: true
          },
          data: {foo: 'foo@bar.com'}
        })
      ).toBe(true);
    });

    it('should return false if not an email.', () => {
      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            email: true
          },
          data: {foo: '123'}
        })
      ).toBe(false);
    });

    // TODO should act like filled (false evals the inverse of the operator)
    // it('should return inverse', () => {
    //   expect(
    //     evalCond({
    //       cond: {
    //         questionId: 'foo',
    //         email: false
    //       },
    //       data: {foo: 'foo@bar.com'}
    //     })
    //   ).toBe(false);
    //
    //   expect(
    //     evalCond({
    //       cond: {
    //         questionId: 'foo',
    //         email: false
    //       },
    //       data: {foo: '123'}
    //     })
    //   ).toBe(true);
    // });
  });

  describe('compare', () => {
    it('should compare one questionId to another questionId', () => {
      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            compare: {
              equals: 'bar'
            }
          },
          data: {
            foo: 50,
            bar: 50
          }
        })
      ).toBe(true);

      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            compare: {
              lessThan: 'bar'
            }
          },
          data: {
            foo: 50,
            bar: 50
          }
        })
      ).toBe(false);
    });

    it('should eval conds in the data', () => {
      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            compare: {
              cond: 'barCond'
            }
          },
          data: {
            foo: 50,
            barCond: {
              questionId: 'foo',
              equals: 50
            }
          }
        })
      ).toBe(true);
    });
  });

  describe('cond', () => {
    it('should return true if foo === "bar"', () => {
      expect(
        evalCond({
          cond: {cond: {questionId: 'foo', equals: 'bar'}},
          data: {foo: 'bar'}
        })
      ).toBe(true);
    });

    it('should return false if foo !== "bar"', () => {
      expect(
        evalCond({
          cond: {cond: {questionId: 'foo', equals: 'bar'}},
          data: {foo: 'baz'}
        })
      ).toBe(false);
    });
  });

  describe('customOperators', () => {
    it('should console.error if missing operator', () => {
      startErrorSupression();
      expect(
        evalCond({
          cond: {questionId: 'foo', barOp: true},
          data: {foo: 'bar'}
        })
      ).toBe(true);
      expect(consoleErrorSpy).toBeCalled();
      expect(consoleErrorSpy.mock.calls).toMatchSnapshot();
      stopErrorSuppression();
    });

    it('should use customOperators', () => {
      startErrorSupression();
      expect(
        evalCond({
          cond: {questionId: 'foo', barOp: true},
          data: {foo: 'baz'},
          customOperators: {
            barOp: ({value, param}) => (param ? value === 'bar' : true)
          }
        })
      ).toBe(false);
      expect(consoleErrorSpy).not.toBeCalled();
      stopErrorSuppression();
    });
  });

  describe('regex', () => {
    it('should return true if nil', () => {
      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            regex: '[a-z]+'
          },
          data: {}
        })
      ).toBe(true);

      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            regex: '[a-z]+'
          },
          data: {foo: null}
        })
      ).toBe(true);
    });

    it('should return true if matches', () => {
      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            regex: '[a-z]+'
          },
          data: {foo: 'abc'}
        })
      ).toBe(true);
    });

    it('should return false if not matches.', () => {
      expect(
        evalCond({
          cond: {
            questionId: 'foo',
            regex: '[a-z]+'
          },
          data: {foo: '123'}
        })
      ).toBe(false);
    });
  });
});

describe('condDependentFields()', () => {
  it('should pull all dependent questionIds', () => {
    const cond = {
      and: [
        {
          questionId: 'one'
        },
        {
          or: [
            {
              questionId: 'two'
            },
            {
              not: {
                questionId: 'three'
              }
            }
          ]
        }
      ]
    };

    const dependentFields = condDependentFields(cond);

    expect(dependentFields).toEqual(['one', 'two', 'three']);
  });
});

describe('evalCondValid()', () => {
  const customFieldTypes = {
    alwaysError: ({field}) => ({
      _genIsValid: () => false,
      name: field.questionId
    }),
    alwaysValid: ({field}) => ({
      _genIsValid: () => true,
      name: field.questionId
    })
  };
  it('should use getFieldErrors() when no operator is specified', () => {
    const fields = [
      {
        type: 'alwaysError',
        questionId: 'alwaysError'
      },
      {
        type: 'alwaysValid',
        questionId: 'alwaysValid'
      }
    ];

    const lookupTable = buildLookupTable({
      fields,
      customFieldTypes
    });

    const alwaysError = evalCondValid({
      cond: {
        questionId: 'alwaysError'
      },
      data: {
        alwaysError: 'foo'
      },
      customFieldTypes,
      lookupTable
    });

    const alwaysValid = evalCondValid({
      cond: {
        questionId: 'alwaysValid'
      },
      data: {
        alwaysValid: 'foo'
      },
      customFieldTypes,
      lookupTable
    });

    expect(alwaysError).toBe(false);
    expect(alwaysValid).toBe(true);
  });
});
