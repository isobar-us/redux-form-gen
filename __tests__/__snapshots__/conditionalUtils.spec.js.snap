// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`evalCond() customOperators should console.error if missing operator 1`] = `
Array [
  Array [
    "[FormGenerator] Unknown conditional operator \\"barOp\\". Condition:",
    Object {
      "barOp": true,
      "questionId": "foo",
    },
  ],
]
`;
